conda_forge_tick/hashing.py

```diff
@@ -0,0 +1,97 @@
+import hashlib
+from multiprocessing import Process, Pipe
+import time
+import math
+
+import requests
+
+
+def _hash_url(url, hash_type, progress=False, conn=None):
+    _hash = None
+    try:
+        ha = getattr(hashlib, hash_type)()
+        resp = requests.get(url, stream=True)
+        if resp.status_code == 200:
+            if 'Content-length' in resp.headers:
+                num = math.ceil(float(resp.headers['Content-length']) / 8192)
+            elif resp.url != url:
+                # redirect for download
+                h = requests.head(resp.url).headers
+                if 'Content-length' in h:
+                    num = math.ceil(float(h['Content-length']) / 8192)
+                else:
+                    num = None
+            else:
+                num = None
+
+            t0 = time.time()
+            loc = 0
+            for itr, chunk in enumerate(resp.iter_content(chunk_size=8192)):
+                ha.update(chunk)
+                if num is not None and int((itr+1) / num * 25) > loc and progress:
+                    eta = (time.time() - t0) / (itr+1) * (num - (itr+1))
+                    loc = int((itr+1) / num * 25)
+                    print("eta % 7.2fs: [%s%s]" % (
+                        eta,
+                        "".join(["=" * loc]),
+                        "".join([" " * (25 - loc)]),
+                        ))
+
+            _hash = ha.hexdigest()
+        else:
+            _hash = None
+    except Exception as e:
+        _hash = (repr(e),)
+    finally:
+        if conn is not None:
+            conn.send(_hash)
+            conn.close()
+        else:
+            return _hash
+
+
+def hash_url(url, timeout=None, progress=False, hash_type='sha256'):
```

Comment:
This seems like it could be one line in xonsh

Reply:
The xonsh version is hanging the bot for more than 10 minutes on gmsh. Idk why but the above works around this with a timeout. Also we had an issue where using urllib versus requests generated different hashes. Conda build uses requests so we have to match that. 

Cc @chrisburr who tracked down the requests versus urllib issue.

Reply:
See also this issue: https://github.com/regro/cf-scripts/issues/500